opensafely/check.py

```diff
@@ -0,0 +1,115 @@
+from opensafely._vendor import requests
+from opensafely._vendor.ruamel.yaml import YAML
+import os
+import re
+import glob
+import subprocess
+
+DESCRIPTION = "Check the opensafely project for correctness"
+
+RESTRICTED_DATASETS = {"icnarc": ["admitted_to_icu"]}
+
+# todo: dummy location and data
+PERMISSIONS_URL = "https://raw.githubusercontent.com/Jongmassey/research_repository_permissions/main/repository_permissions.yaml"
+
+
+def add_arguments(parser):
+    pass
+
+
+def main():
+    repo_name = get_repository_name()
+    permissions = get_datasource_permissions(PERMISSIONS_URL)
+    allowed_datasets = get_allowed_datasets(repo_name, permissions)
+    datasets_to_check = {
+        k: v
+        for k, v in RESTRICTED_DATASETS.items()
+        if k not in allowed_datasets
+    }
+    files_to_check = get_local_py_files()
+
+    found_datasets = {
+        dataset: check_dataset(functions, files_to_check)
+        for dataset, functions in datasets_to_check.items()
+        if check_dataset(functions, files_to_check)
+    }
+
+    if found_datasets:
+        exit(print_violations(found_datasets))
+    else:
+        print("Success")
+
+
+def print_violations(found_datasets):
+    s = "Usage of restricted datasets found:\n"
+    for d, functions in found_datasets.items():
+        s = s + (f"{d}:\n")
+        for fn, files in functions.items():
+            s = s + (f"- {fn}\n")
+            for f, lines in files.items():
+                s = s + (f"  - {f}:\n")
+                for ln, line in lines.items():
+                    s = s + (f"    line {ln}: {line}\n")
+    return s
+
+
+def check_dataset(functions, files_to_check):
+    found_functions = {}
+    for function in functions:
+        regex = re.compile(rf"\.{function}\(")
+        found_files = {}
+        for f in files_to_check:
+            matches = check_file(f, regex)
+            if matches:
+                found_files[f] = matches
+        if found_files:
+            found_functions[function] = found_files
+    return found_functions
+
+
+def check_file(filename, regex):
+    found_lines = {}
+    with open(filename, "r") as f:
+        for ln, line in enumerate(f.readlines(), start=1):
+            if regex.search(line) and not line.lstrip().startswith("#"):
```

Comment:
nit: I would write this like so:

```python
if line.lstrip().startswith("#"):
    continue
if regex.search(line):
    ...
```
My thinking is that a) logicially, its just skipping comment lines, which feels right to me b) it avoids running the regex on comment lines.